espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ActivityHelper.java

```diff
@@ -32,4 +56,42 @@ public static Activity getCurrentActivity() throws AppiumException {
 
         throw new AppiumException("Failed to get current Activity");
     }
+
+    private static String getFullyQualifiedActivityName(Instrumentation mInstrumentation, String appActivity) {
+        Context context = mInstrumentation.getTargetContext();
+        // If it's not fully qualified, make it fully qualified
+        return appActivity.startsWith(".")
+                ? context.getPackageName() + appActivity
+                : appActivity;
+    }
+
+    public static void startActivityViaInstruments(String appActivity, boolean waitForActivity) {
+        logger.info(String.format("Starting activity '%s'", appActivity));
+        Instrumentation mInstrumentation = InstrumentationRegistry.getInstrumentation();
+        Instrumentation.ActivityMonitor mSessionMonitor = mInstrumentation.addMonitor(appActivity, null, false);
+        Intent intent = new Intent(Intent.ACTION_MAIN);
+        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+        String fullyQualifiedAppActivity = getFullyQualifiedActivityName(mInstrumentation, appActivity);
+        intent.setClassName(mInstrumentation.getTargetContext(), fullyQualifiedAppActivity);
+        mInstrumentation.startActivitySync(intent);
+        if (waitForActivity) {
+            Activity mCurrentActivity = mInstrumentation.waitForMonitor(mSessionMonitor);
+            logger.info(String.format("Activity '%s' started", mCurrentActivity.getLocalClassName()));
+        }
+    }
+
+    public static void startActivityViaScenario(String activity) throws AppiumException {
+        logger.info(String.format("Starting activity '%s'", activity));
+        Instrumentation mInstrumentation = InstrumentationRegistry.getInstrumentation();
+        final Class<? extends Activity> activityClass;
+        try {
+            //noinspection unchecked
+            activityClass = (Class<? extends Activity>) Class
+                    .forName(getFullyQualifiedActivityName(mInstrumentation, activity));
+        } catch (ClassNotFoundException | ClassCastException e) {
+            throw new AppiumException(e);
+        }
+        ActivityScenario.launch(activityClass);
```

Comment:
Below error happened here when I invoked  https://github.com/appium/ruby_lib_core/blob/d104fc91e849dc183267ecd3a4074e9fa1c94a38/test/functional/android/android/device_test.rb#L129

```
 android.content.ActivityNotFoundException: Unable to find explicit activity class {io.appium.espressoserver.test/androidx.test.core.app.InstrumentationActivityInvoker$BootstrapActivity}; have you declared this activity in your AndroidManifest.xml?
        at android.app.Instrumentation.checkStartActivityResult(Instrumentation.java:1933)
        at android.app.Instrumentation.execStartActivity(Instrumentation.java:1616)
        at androidx.test.runner.MonitoringInstrumentation.execStartActivity(MonitoringInstrumentation.java:508)
        at android.app.ContextImpl.startActivity(ContextImpl.java:878)
        at android.app.ContextImpl.startActivity(ContextImpl.java:855)
        at androidx.test.core.app.InstrumentationActivityInvoker.startActivity(InstrumentationActivityInvoker.java:352)
        at androidx.test.core.app.ActivityScenario.launch(ActivityScenario.java:207)
        at androidx.test.core.app.ActivityScenario.launch(ActivityScenario.java:184)
        at io.appium.espressoserver.lib.helpers.ActivityHelper.startActivityViaScenario(ActivityHelper.java:96)
        at io.appium.espressoserver.lib.handlers.StartActivity.handle(StartActivity.java:28)
        at io.appium.espressoserver.lib.handlers.StartActivity.handle(StartActivity.java:24)
        at io.appium.espressoserver.lib.http.Router.route(Router.java:311)
        at io.appium.espressoserver.lib.http.Server.serve(Server.java:68)
        at fi.iki.elonen.NanoHTTPD$HTTPSession.execute(NanoHTTPD.java:945)
        at fi.iki.elonen.NanoHTTPD$ClientHandler.run(NanoHTTPD.java:192)
        at java.lang.Thread.run(Thread.java:764)
from UnknownError: An unknown server-side error occurred while processing the command. Original error: android.content.ActivityNotFoundException: Unable to find explicit activity class {io.appium.espressoserver.test/androidx.test.core.app.InstrumentationActivityInvoker$BootstrapActivity}; have you declared this activity in your AndroidManifest.xml?
```

The `activityClass` was `io.appium.android.apis.accessibility.AccessibilityNodeProviderActivity`. The activity had intent filter as below. The test app was https://github.com/appium/ruby_lib_core/blob/d104fc91e849dc183267ecd3a4074e9fa1c94a38/test/functional/app/api.apk.zip which is usually we test with it.

```
        <activity
            android:label="@ref/0x7f0c0399"
            android:name=".accessibility.AccessibilityNodeProviderActivity"
            android:enabled="@ref/0x7f090002">

            <intent-filter>

                <action
                    android:name="android.intent.action.MAIN" />

                <category
                    android:name="appium.android.intent.category.SAMPLE_CODE" />
            </intent-filter>
        </activity>
```

Reply:
@KazuCocoa can you please try the current commit?

Reply:
@mykola-mokhnach I tried your fork and replaced test apk but still see the same failures, unfortunately. https://gist.github.com/SrinivasanTarget/95abc4bb8d33c470a289a7d4ae131226

Any thoughts?

Reply:
Below is the intent passed to the invoker's `startActivity`
```
Intent { act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] cmp=io.appium.android.apis/.accessibility.AccessibilityNodeProviderActivity }
```

Reply:
Below is the startActivity in invoker. `getIntentForActivity(BootstrapActivity.class)` is the cause of the error, it looks

```
  @Override
  public void startActivity(Intent intent) {
    // Close empty activities and bootstrap activity if it's running. This might happen if the
    // previous test crashes before it cleans up the state.
    getTargetContext().sendBroadcast(new Intent(FINISH_BOOTSTRAP_ACTIVITY));
    getTargetContext().sendBroadcast(new Intent(FINISH_EMPTY_ACTIVITIES));

    activityResultWaiter = new ActivityResultWaiter(getTargetContext());

    // Note: Instrumentation.startActivitySync(Intent) cannot be used here because BootstrapActivity
    // may start in different process. Also, we use PendingIntent because the target activity may
    // set "exported" attribute to false so that it prohibits starting the activity outside of their
    // package. With PendingIntent we delegate the authority to BootstrapActivity.
    getTargetContext()
        .startActivity(
            getIntentForActivity(BootstrapActivity.class)
                .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK)
                .putExtra(
                    TARGET_ACTIVITY_INTENT_KEY,
                    PendingIntent.getActivity(
                        getTargetContext(),
                        /*requestCode=*/ 0,
                        intent,
                        /*flags=*/ PendingIntent.FLAG_UPDATE_CURRENT)));
  }
```

Reply:
I tried to replace this method to call `startActivityViaInstruments` once. Then, I could launch the activity. Could call some methods after this.

mmm, `ActivityScenario` is not much stable yet??

Reply:
might be. Let me try ActivityRule instead

Reply:
the current commit uses TestActivityRule. Would  you like to try that one?

Reply:
Almost worked. But failed to launch 


```
# ok
@driver.start_activity app_package: 'io.appium.android.apis', app_activity: '.accessibility.AccessibilityNodeProviderActivity'

# never back (and failed to launch error happened)
@driver.start_activity app_package: 'io.appium.android.apis', app_activity: '.ApiDemos'
```

The 2nd one is just an activity specified as `appActivity` in the capability.
I am not sure, but if a user try to launch an activity which launched by 

```
Selenium::WebDriver::Error::UnknownError: An unknown server-side error occurred while processing the command. Original error: java.lang.RuntimeException: Could not launch intent Intent { act=android.intent.action.MAIN flg=0x10000000 cmp=io.appium.android.apis/.accessibility.AccessibilityNodeProviderActivity } within 45 seconds. Perhaps the main thread has not gone idle within a reasonable amount of time? There could be an animation or something constantly repainting the screen. Or the activity is doing network calls on creation? See the threaddump logs. For your reference the last time the event queue was idle before your activity launch request was 1548208999277 and now the last time the queue went idle was: 1548208999277. If these numbers are the same your activity might be hogging the event queue.
        at androidx.test.runner.MonitoringInstrumentation.startActivitySync(MonitoringInstrumentation.java:453)
        at androidx.test.rule.ActivityTestRule.launchActivity(ActivityTestRule.java:358)
        at io.appium.espressoserver.lib.helpers.ActivityHelper.startActivityViaTestRule(ActivityHelper.java:98)
        at io.appium.espressoserver.lib.handlers.StartActivity.handle(StartActivity.java:28)
        at io.appium.espressoserver.lib.handlers.StartActivity.handle(StartActivity.java:24)
        at io.appium.espressoserver.lib.http.Router.route(Router.java:311)
        at io.appium.espressoserver.lib.http.Server.serve(Server.java:68)
        at fi.iki.elonen.NanoHTTPD$HTTPSession.execute(NanoHTTPD.java:945)
        at fi.iki.elonen.NanoHTTPD$ClientHandler.run(NanoHTTPD.java:192)
        at java.lang.Thread.run(Thread.java:764)
from UnknownError: An unknown server-side error occurred while processing the command. Original error: java.lang.RuntimeException: Could not launch intent Intent { act=android.intent.action.MAIN flg=0x10000000 cmp=io.appium.android.apis/.accessibility.AccessibilityNodeProviderActivity } within 45 seconds. Perhaps the main thread has not gone idle within a reasonable amount of time? There could be an animation or something constantly repainting the screen. Or the activity is doing network calls on creation? See the threaddump logs. For your reference the last time the event queue was idle before your activity launch request was 1548208999277 and now the last time the queue went idle was: 1548208999277. If these numbers are the same your activity might be hogging the event queue.

```

> For your reference the last time the event queue was idle before your activity launch request was 1548208999277 and now the last time the queue went idle was: 1548208999277. If these numbers are the same your activity might be hogging the event queue.

mmm, we probably should handle `instrumentation.waitForMonitor(mSessionMonitor);`..?

Reply:
ok, i don't have any other ideas except for the recent commit.

Reply:
Thanks! I also have no good idea...

Can you handle `appWaitActivity = ""` as same as `null` case?
I forget the reason, but Ruby binding send `""` by default (maybe, so long time ago's some workaround... Then, `fullyQualifiedWaitActivity` will be `""` and never response back.
e.g.: `{"appPackage":"io.appium.android.apis","appActivity":".accessibility.AccessibilityNodeProviderActivity","appWaitPackage":"","appWaitActivity":""}`

I will also stop sending blank

----

Except for it, worked well on my env. Thanks for much!


If we get `*` request, we consider `*` case in the futuer.
http://appium.io/docs/en/writing-running-appium/caps/